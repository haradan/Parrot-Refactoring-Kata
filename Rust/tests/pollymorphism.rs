use insta::assert_snapshot;
use itertools::Itertools;
use pollymorphism::{AfricanParrot, EuropeanParrot, NorwegianBlue, Parrot};

#[test]
fn parrot_speeds() {
    let parrots: &[(&str, Box<dyn Parrot>)] = &[
        ("European parrot", Box::new(EuropeanParrot {})),
        (
            "Unladen african parrot",
            Box::new(AfricanParrot {
                number_of_coconuts: 0,
            }),
        ),
        (
            "African parrot with 1 coconut",
            Box::new(AfricanParrot {
                number_of_coconuts: 1,
            }),
        ),
        (
            "African parrot with 2 coconuts",
            Box::new(AfricanParrot {
                number_of_coconuts: 2,
            }),
        ),
        (
            "Norwegian blue",
            Box::new(NorwegianBlue {
                voltage: 0.0,
                nailed: false,
            }),
        ),
        (
            "1V Norwegian blue",
            Box::new(NorwegianBlue {
                voltage: 1.0,
                nailed: false,
            }),
        ),
        (
            "1.5V Norwegian blue",
            Box::new(NorwegianBlue {
                voltage: 1.5,
                nailed: false,
            }),
        ),
        (
            "4V Norwegian blue",
            Box::new(NorwegianBlue {
                voltage: 4.0,
                nailed: false,
            }),
        ),
        (
            "4V nailed Norwegian blue",
            Box::new(NorwegianBlue {
                voltage: 4.0,
                nailed: true,
            }),
        ),
    ];
    let parrot_speeds: String = parrots
        .iter()
        .map(|tuple| format!("{:>31}: {:>2}", tuple.0, tuple.1.speed().unwrap()))
        .join("\n");
    assert_snapshot!(parrot_speeds);
}
