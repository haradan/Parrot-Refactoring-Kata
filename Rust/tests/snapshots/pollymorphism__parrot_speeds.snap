---
source: tests/pollymorphism.rs
expression: parrot_speeds
---
                European parrot: 12
         Unladen african parrot: 12
  African parrot with 1 coconut:  3
 African parrot with 2 coconuts:  0
                 Norwegian blue:  0
              1V Norwegian blue: 12
            1.5V Norwegian blue: 18
              4V Norwegian blue: 24
       4V nailed Norwegian blue:  0
